examples/PPO/train.py

```diff
@@ -123,33 +149,25 @@ def main():
     test_flag = 0
     total_steps = 0
     while total_steps < args.train_total_steps:
-        obs, actions, rewards = collect_trajectories(
+        trajectories = collect_trajectories(
             env, agent, scaler, episodes=args.episodes_per_batch)
-        total_steps += obs.shape[0]
-
-        pred_values = agent.value_predict(obs)
-
-        # scale rewards
-        scale_rewards = rewards * (1 - args.gamma)
-
-        discount_sum_rewards = calc_discount_sum_rewards(
-            scale_rewards, args.gamma)
-        discount_sum_rewards = discount_sum_rewards.astype('float32')
+        total_steps += sum([t['obs'].shape[0] for t in trajectories])
+        total_train_rewards = sum([np.sum(t['rewards']) for t in trajectories])
 
-        advantages = calc_gae(scale_rewards, pred_values, args.gamma, args.lam)
-        # normalize advantages
-        advantages = (advantages - advantages.mean()) / (
-            advantages.std() + 1e-6)
-        advantages = advantages.astype('float32')
+        train_obs, train_actions, train_advantages, train_discount_sum_rewards = build_train_data(
+            trajectories, agent)
 
-        policy_loss, kl = agent.policy_learn(obs, actions, advantages)
-        value_loss = agent.value_learn(obs, discount_sum_rewards)
+        policy_loss, kl = agent.policy_learn(
+            np.concatenate(train_obs), np.concatenate(train_actions),
```

Comment:
It would be better to move data processing operations such as `concatenate` inside function `build_train_data`. Also, you have concatenated the same array for several times, e.g, `train_obs`, these repeated operations are wasteful use of computation.